### PR TITLE
Make the inserter behave as a popover

### LINKED ISSUE #22858: Regression: the new Block Inserter doesn't constrain tabbing
**Describe the bug**
The new Block Inserter doesn't constrain tabbing with the Tab key.

In the old one, tabbing was constrained. Although we're not treating the Block Inserter as a modal, I think the previous behavior was beneficial. Not sure about the reasoning behind the removal of the tabbing constraining. I'd appreciate anyone pointing me to the discussion where this behavior removal was considered.

**To reproduce**
- test on default WordPress 5.4 block editor first
- open the main block inserter:

Screenshot 2020-06-03 at 15 59 24

- focus is moved within the inserter 
- use the Tab key to navigate the inserter content
- see that tabbing is constrained within the inserter: once you reach the last focusable element, focus is moved to the beginning and vice-versa
- activate the Gutenberg plugin, right now it's version 8.2.1 
- open the main block inserter: you get the new one:
Screenshot 2020-06-03 at 15 58 24

- use the Tab key to navigate the inserter content
- by default the inserter displays 70 blocks so you will need to press Tab 70 times
- I'll refrain from expressing usability concerns about the removal of the expandable block groups: having to tab dozens of times, or scroll through a long list of blocks, doesn't seem a great improvement to me
- once you reach the last focusable element, tabbing is not constrained within the inserter any longer (and vice-versa)
- keeping tabbing exits the inserter and moves focus to the rest of the UI, which isn't ideal

**Expected behavior**
Tabbing to be constrained.

**Editor version (please complete the following information):**
- WordPress version: 5.4
- Gutenberg plugin 8.2.1

### PR DESCRIPTION
**Pull request by youknowriad**
closes #22858 

I've been thinking about the behavior of the inserter and here's my findings so far. If we want to make the collapsible panel accessible properly It means:

 - Opening means moves the focus inside the panel
 - tabbing is constrained.
 - Constrained tabbing like suggested in #22858 is not sufficient though, moving the focus outside the panel should close the panel. It happens when inserting a block (focus moved to content) or when clicking “Escape”. If we don’t close the panel when we insert blocks and at the same time move the focus out of the panel, we’re going to create focus traps because you’ll be able to move back to the panel using tabs and stay trapped. (See https://www.w3.org/TR/UNDERSTANDING-WCAG20/keyboard-operation-trapping.html )

That said this solution gives away the possibility to insert multiple blocks and patterns in a row. So we have two options we can explore later:

 - Add a modifier key,  when you hold it and insert a block, it keeps the inserter open and doesn’t move the focus to the content area which means the inserter stays open. (This allows the multi-insertion to work both with keyboard and mouse while right now it only works for mouse users)
 - Don't focus the content when inserting patterns as it's not clear what block in the pattern should be focused.
 - By default don’t focus on the content when you insert a block, unless you hit “Escape” to exit the insert.

### CODE REVIEW

File: packages/edit-post/src/components/layout/popover-wrapper.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+			return this.props.children;
+		}
+	}
+);
+
+const FocusManaged = withConstrainedTabbing(
+	withFocusReturn( ( { children } ) => children )
+);
+
+export default function PopoverWrapper( { onClose, children } ) {
```
Comment (youknowriad) [inline packages/edit-post/src/components/layout/popover-wrapper.js:32]: This was extracted as is from Popover, ideally, it should be a reusable hook/component (or multiple).

**Review by jasmussen — APPROVED** [review]
This is what I see:

popoverinserter

This fixes important issues, and while it would be good to go back and revisit some of the specific behaviors, it seems good to get this in.

**Comment by enriquesanchez** [comment]
I think this is a great improvement, thanks Riad! Focus is constrained and predictable. Hitting `esc` takes me back to where I was 👍 

Here's a gif of my test:

inserter


You'll notice that the editor's content scrolls along with my position in the inserter as I tab. I'm not sure this should happen. What I find curious is that I don't see that happening on Joen's test. Maybe something funny happening on my local install?

I also think screen-reader navigation would be much friendlier if we announced each group the user lands on while tabbing. Right now I only hear "Image text, Buttons text, Shortcode text", so it's hard to understand that you're on the first item of a group. I'll create a different issue as this is out of scope for this one but still wanted to note it.

**Comment by jasmussen** [comment]
A small thing regressed, perhaps in a rebase, from the time I approved this. Two instances of `height: 100%` are missing:

Screenshot 2020-08-10 at 08 49 13

When you add those back in the inspector, things work as they should. Without them, we have the double scrollbar issue again.

**Comment by StevenDufresne** [comment]
@youknowriad When you click inside a block, the main inserter closes. If you have a block with a template lock and a select list of allowed blocks, how do you add those children blocks?

**Comment by youknowriad** [comment]
>  If you have a block with a template lock and a select list of allowed blocks, how do you add those children blocks?

 Not sure I understand what you mean here? It just works like previous versions. You click the appender of that container block and add the children blocks?

**Comment by StevenDufresne** [comment]
> > If you have a block with a template lock and a select list of allowed blocks, how do you add those children blocks?
> 
> Not sure I understand what you mean here? It just works like previous versions. You click the appender of that container block and add the children blocks?

Given this block: https://gist.github.com/ryelle/cacdba5c888153d97eae38db775dec7f (maybe there's an issue with this block?)

The appender does not show up. The appender visible in the screenshot is not the containers appender.

**Comment by StevenDufresne** [comment]
Actually, that's a bad example, it has a template lock.

**Comment by apeatling** [comment]
This change is quite a big regression for building pages using block patterns. I would not expect the inserter to close when I select a pattern, and I expect users would want to add multiple patterns in one go. Also, I lose my scroll position, so when reopening I have to find where I was again.

**Comment by shaunandrews** [comment]
This PR has made it incredibly difficult to add multiple blocks or patterns to a document in the post editor. It _also_ makes it difficult to add multiple blocks or widgets to a sidebar with the widget editor. [See #26223]

> By default don’t focus on the content when you insert a block, unless you hit “Escape” to exit the insert.

I think its time we do this.

I could see the same behavior for the other sidebars as well; Opening a sidebar would trap focus. Actions within the sidebar _can_ move focus away, and you could still move focus back into the sidebar — but you'd be able to use `esc` to close the sidebar and remove focus.

**Comment by paaljoachim** [comment]
Hey Shaun

How do you suggest we move this onward?

**Comment by shaunandrews** [comment]
There's a PR that revisits this behavior over here: https://github.com/WordPress/gutenberg/pull/28191

**Comment by paaljoachim** [comment]
Thanks!

**Pull request merged**